### PR TITLE
use symfony http client

### LINKED ISSUE #6: The Publisher class does not work with a self-signed certificate
Hello,

For one of my projects, I installed Mercure in HTTPS mode and so in locally my Mercure hub is accessible via a self signed certificate.  
Except that the `publish` function does not take into account this case.  
The function could look like this:

```php
$result = @file_get_contents($this->hubUrl, false, stream_context_create(
    [
        'http' => [
            'method' => 'POST',
            'header' => "Content-type: application/x-www-form-urlencoded\r\nAuthorization: Bearer $jwt",
            'content' => $postData,
        ],
       'ssl' => [
           'verify_peer' => false,
           'verify_peer_name'=> false,
           'allow_self_signed' => true
        ]
    ]
));
```
Providing a configuration in the Mercury bundle can be a good idea too.
If this issue seems relevant, I want to try to contribute for the first time to Symfony :)

### LINKED ISSUE #7: Publisher throw error "failed to open stream: operation failed"
Using the publisher or the following code, same as used in the [Publisher class](https://github.com/symfony/mercure/blob/master/src/Publisher.php).
I got the error `failed to open stream: operation failed`

````php
$url = 'https://..../hub';
$jwt = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJtZXJjdXJlIjp7InN1YnNjcmliZSI6W10sInB1Ymxpc2giOltdfX0.bxavdlHOyum1AbQteaEdtJuEU0HyA2pp_s2b1HqNlN0';
$postData = 'topic=hello&data=welcome';

$result = @file_get_contents(
    $url,
    false,
    stream_context_create(
        [
            'http' => [
                'method' => 'POST',
                'header' => "Content-type: application/x-www-form-urlencoded\r\nAuthorization: Bearer $jwt",
                'content' => $postData,
            ],
        ]
    )
);

print_r($result); // null
print_r(error_get_last()['message']); //file_get_contents(https://...): failed to open stream: operation failed
exit;
````

But when use my own publisher using curl works fine:

````php
$ch = curl_init($url);
curl_setopt_array(
    $ch,
    [
        CURLOPT_URL => $url,
        CURLOPT_POST => 1,
        CURLOPT_POSTFIELDS => $postData,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => [
            'Content-type: application/x-www-form-urlencoded',
            "Authorization: Bearer $jwt",
        ],
    ]
);

$server_output = curl_exec($ch);
curl_close($ch);
print_r($server_output);//e6dab35e-5158-416a-8613-b8023936bce1
exit;
````
I'm not sure what's the problem :X, but why not use curl in the publisher by default to do this?

### PR DESCRIPTION
**Pull request by azjezz**
fixes #6 
fixes #7

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -16,7 +16,8 @@
         }
     ],
     "require": {
-        "php": "^7.1.3"
+        "php": "^7.1.3",
+        "symfony/http-client": "dev-master"
```
Comment (dunglas) [inline composer.json]:
```suggestion
        "symfony/http-client": "^4.3@beta"
```

File: tests/PublisherTest.php [hunk 2]
```diff
@@ -24,20 +27,22 @@ class PublisherTest extends TestCase
 {
     const URL = 'https://demo.mercure.rocks/hub';
     const JWT = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJtZXJjdXJlIjp7InN1YnNjcmliZSI6WyJmb28iLCJiYXIiXSwicHVibGlzaCI6WyJmb28iXX19.LRLvirgONK13JgacQ_VbcjySbVhkSmHy3IznH3tA9PM';
+    const AUTH_HEADER = 'Bearer ' . self::JWT;
```
Comment (dunglas) [inline tests/PublisherTest.php]:
```suggestion
    const AUTH_HEADER = 'Bearer '.self::JWT;
```

**Comment by azjezz** [comment]
:+1:

**Comment by jevillard** [comment]
Good work ! I think adding an array of options in the request would be interesting to allow developers to add the options they want to the HTTP request 😉

**Comment by nicolas-grekas** [comment]
Adding options is also possible by configuring the http client instance with default ones, if that covers the use case you have in mind.

**Comment by jevillard** [comment]
Ok, I did not remember that possibility, so it's perfect.

File: composer.json [hunk 3]
```diff
@@ -16,7 +16,8 @@
         }
     ],
     "require": {
-        "php": "^7.1.3"
+        "php": "^7.1.3",
+        "symfony/http-client": "^4.3@beta"
```
Comment (fabpot) [inline composer.json]: Could now be changed to `^4.3`

Comment (azjezz) [reply composer.json]: :+1:

File: src/Publisher.php [hunk 4]
```diff
@@ -29,14 +32,14 @@ final class Publisher
     private $httpClient;
 
     /**
-     * @param callable(): string                            $jwtProvider
-     * @param null|callable(string, string, string): string $httpClient
+     * @param callable(): string       $jwtProvider
+     * @param null|HttpClientInterface $httpClient
```
Comment (dunglas) [inline src/Publisher.php]: Can be removed now.

Comment (azjezz) [reply src/Publisher.php]: I thought Symfony cs enforced this ?

**Comment by azjezz** [comment]
cc @dunglas travis-ci is happy :smile:

**Comment by dunglas** [comment]
Thanks @azjezz 🎆

**Pull request merged**